src/Microsoft.ML.Dnn/ImageClassificationTransform.cs

```diff
@@ -1166,7 +1166,7 @@ internal ImageClassificationEstimator(IHostEnvironment env, Options options, Dnn
             _options = options;
             _dnnModel = dnnModel;
             _tfInputTypes = new[] { TF_DataType.TF_STRING };
-            _outputTypes = new[] { new VectorDataViewType(NumberDataViewType.Single), NumberDataViewType.UInt32.GetItemType() };
+            _outputTypes = new DataViewType[] { new VectorDataViewType(NumberDataViewType.Single), new KeyDataViewType(typeof(uint), 5) };
```

Comment:
>_outputTypes [](start = 12, length = 12)

`_outputTypes` doesn't need to be an array of `DataViewType`. All the estimator needs to know is whether the length of the score vector is variable or not (seems to me that it is typically not variable, or at least this information can be inferred from the `DnnModel`).
If you get rid of this field, you will not have to guess the size of the key (which is also not needed by the estimator).

Reply:
This is a good catch, thank you! Quick question, the Image Classification example ResnetV2101TransferLearningTrainTestSplit.cs runs perfectly well when I don't further define _outputTypes, i,e, delete line 69. Does this also mean than the estimator isn't using this field?

Reply:
Actually, I've found that the DataviewType field is needed for the pipeline in ResnetV2101TransferLearningTrainTestSplit.cs to fit properly.

Reply:
It was giving a different error when I removed _outputTypes, but as I now see that was not the bottleneck problem I was having while implementing this KeyType solution.

Reply:
The field `_outputTypes` can be deleted. The only place where it is used is in line 1281 where you have `_outputTypes[0].GetItemType()`. Instead of that you can write `NumberDataViewType.Single`, and then the field is not needed.




---
In reply to: [326229209](https://github.com/dotnet/machinelearning/pull/4228#discussion_r326229209) [](ancestors = 326229209)

Reply:
I missed this, thank you for the catch!